### PR TITLE
feat: add automated version tagging for CI/CD pipeline

### PR DESCRIPTION
**Pull request by hiroyukikumazawa**
#### Overview

This PR introduces a new GitHub Actions workflow, `common_tagging.yml`, that automates the tagging process for our CI/CD pipeline. The automated version tagging ensures consistent version management and supports changelog integration across different components of the Compute Horde monorepo. The tagging format follows `<year>.<month>.<day>-<index>`, incrementing if multiple releases occur on the same day.

#### Key Changes

- **New Workflow**: `common_tagging.yml` added to handle common tagging tasks.
  - **Inputs**:
    - `DIST_VERSION_PREFIX`: Prefix for the distribution (e.g., `executor-preprod`).
  - **Jobs**:
    - Determines the next tag based on the current date and existing tags.
    - Creates and pushes the next tag to the repository.

- **Updates to Existing Workflows**:
  - Workflows such as `executor_cd_preprod.yml` have been updated to utilize the new tagging workflow by calling `common_tagging.yml`.

#### Important Note: `ACTIONS_PAT` Secret

The `ACTIONS_PAT` secret is **critical** for the new tagging workflow to function correctly. It is a Personal Access Token (PAT) that allows the GitHub Actions runner to:
- Fetch all tags from the repository.
- Create and push new tags back to the repository.

##### Setting Up `ACTIONS_PAT`

1. **Generate a Personal Access Token (PAT):**
   - Go to your GitHub account's [Developer settings/Tokens (classic)](https://github.com/settings/tokens).
   - Click **"Generate new token (classic)"** and select the necessary permissions:
     - `repo` (Full control of private repositories).
     - `workflow` (Update GitHub Actions workflows).
   - Copy the generated token; **it will not be shown again.**

2. **Add the `ACTIONS_PAT` Secret to the Repository:**
   - Navigate to the repository's **Settings** > **Secrets and variables** > **Actions**.
   - Click **"New repository secret"**.
   - Name the secret `ACTIONS_PAT` and paste the copied token.

### CODE REVIEW

**Comment by hiroyukikumazawa** [comment]
@ppolewicz 

In the tagging workflow, I have replaced the use of the personal access token (`ACTIONS_PAT`) with the GitHub-provided `GITHUB_TOKEN`.

**Important Note:**
To ensure the updated workflow functions correctly, we need to adjust the **Workflow permissions** settings in the repository's Actions settings. Please follow these steps:

1. Go to **Settings** > **Actions** > **General**.
2. Under **Workflow permissions**, ensure that "Read and write permissions" is selected.

File: .github/workflows/common_tagging.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+on:
+  workflow_call:
+    inputs:
+      DIST_VERSION_PREFIX:
+        required: true
+        type: string
+
+env:
+  CHANGE_LOG_BRANCH: "staging"
```
Comment (mzukowski-reef) [inline .github/workflows/common_tagging.yml:11]: unused? remove?

**Review by mzukowski-reef — CHANGES_REQUESTED** [review]

File: .github/workflows/common_tagging.yml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+          LATEST_TAG=$(git tag -l "${{ inputs.DIST_VERSION_PREFIX }}-${DATE_TAG}-*" | sort -V | tail -n 1)
+
+          if [ -z "$LATEST_TAG" ]; then
+            NEXT_TAG="${DATE_TAG}-1"
+          else
+            COUNTER=$(echo "$LATEST_TAG" | awk -F '-' '{print $NF}')
+            NEXT_TAG="${DATE_TAG}-$((COUNTER + 1))"
+          fi
+
+          echo "NEXT_TAG=$NEXT_TAG" >> $GITHUB_ENV
```
[the remark this answers is not in the document]
Comment (hiroyukikumazawa) [reply .github/workflows/common_tagging.yml:41]: good point!

File: pyproject.toml [hunk 3]
```diff
@@ -12,4 +12,34 @@ dev = [
     "-e file:///${PROJECT_ROOT}/miner",
     "-e file:///${PROJECT_ROOT}/executor",
     "-e file:///${PROJECT_ROOT}/validator",
-]
\ No newline at end of file
+]
+
+[tool.towncrier]
```
[the remark this answers is not in the document]
Comment (hiroyukikumazawa) [reply pyproject.toml:17]: atm, the towncrier part is not used due to the conflict
but it's useful in the future

[the remark this answers is not in the document]
Comment (hiroyukikumazawa) [reply pyproject.toml:17]: @mzukowski-reef what do you think?

**Review by mzukowski-reef — APPROVED** [review]

**Pull request closed without merging**